web/src/bot_data.ts

```diff
@@ -0,0 +1,133 @@
+import _ from "lodash";
+
+import * as people from "./people";
+
+export type Service = {
+    base_url: string;
+    interface: number;
+    config_data: {key: string};
+    service_name: string;
+    token: string;
+};
+
+export type CleanBot = {
+    api_key: string;
+    avatar_url: string;
+    bot_type: number;
+    default_all_public_streams: boolean;
+    default_events_register_stream: string | undefined;
+    default_sending_stream: string | undefined;
+    email: string;
+    full_name: string;
+    is_active: boolean;
+    owner: string;
+    owner_id: number;
+    user_id: number;
+};
+
+export type Bot = CleanBot & {
+    services: Service[];
+    extra?: string;
+};
+
+export type BotDataParams = {
+    realm_bots: Bot[];
+};
+
+const bots = new Map<number, CleanBot>();
+
+const bot_fields = [
+    "api_key",
+    "avatar_url",
+    "bot_type",
+    "default_all_public_streams",
+    "default_events_register_stream",
+    "default_sending_stream",
+    "email",
+    "full_name",
+    "is_active",
+    "owner", // TODO: eliminate
+    "owner_id",
+    "user_id",
+];
+
+const services = new Map<number, Service[]>();
+const services_fields = ["base_url", "interface", "config_data", "service_name", "token"];
+
+export function all_user_ids(): number[] {
+    return [...bots.keys()];
+}
+
+export function add(bot: Bot): void {
+    const clean_bot = _.pick(bot, bot_fields) as CleanBot;
+    bots.set(bot.user_id, clean_bot);
+    const clean_services = bot.services.map(
+        (service) => _.pick(service, services_fields) as Service,
+    );
+    services.set(bot.user_id, clean_services);
+}
+
+export function deactivate(bot_id: number): void {
+    bots.get(bot_id)!.is_active = false;
+}
+
+export function del(bot_id: number): void {
+    bots.delete(bot_id);
+    services.delete(bot_id);
+}
+
+export function update(bot_id: number, bot_update: Bot): void {
```

Comment:
`bot_update` should be unknown.

Reply:
`Partial<Bot>` is more informative.

Reply:
Addressed, thanks!

Reply:
I narrowed it down to `ServerUpdateBotData` with the help of `zerver/openapi/zulip.yaml`. It seems that `update` operation carries `BasicBot` data so I created this type. Yaml description:

```yaml
                            - type: object
                              description: |
                                Event sent to users who can administer a bot user when the bot is
                                configured. Clients may also receive a `realm_user` event that
                                for changes in public data about the bot (name, etc.).

                                The `realm_user` events are sufficient for clients that
                                only need to interact with the bot; this `realm_bot` event
                                type is relevant only for administering bots.

                                Only organization administrators and the user who owns the bot will
                                receive this event.
                              additionalProperties: false
                              properties:
                                id:
                                  $ref: "#/components/schemas/EventIdSchema"
                                type:
                                  allOf:
                                    - $ref: "#/components/schemas/EventTypeSchema"
                                    - enum:
                                        - realm_bot
                                op:
                                  type: string
                                  enum:
                                    - update
                                bot:
                                  allOf:
                                    - description: |
                                        Object containing details about the changed bot.
                                        It contains two properties: the user ID of the bot and
                                        the property to be changed. The changed property is one
                                        of the remaining properties listed below.
                                    - $ref: "#/components/schemas/BasicBot"
                              example:
                                {
                                  "type": "realm_bot",
                                  "op": "update",
                                  "bot":
                                    {
                                      "user_id": 37,
                                      "services":
                                        [
                                          {
                                            "base_url": "http://hostname.domain2.com",
                                            "interface": 2,
                                            "token": "grr8I2APXRmVL0FRTMRYAE4DRPQ5Wlaw",
                                          },
                                        ],
                                    },
                                  "id": 0,
                                }
```